crates/dev-geth/src/main.rs

```diff
@@ -10,96 +10,138 @@
 //! `eth_createAccessList` endpoint. When hardhat adds support for
 //! `eth_createAccessList`, this can be removed from our infrastructure.
 //!
+//! NOTE: This program should only be ran from docker via Dockerfile.dev-geth!
+//! It makes assumptions about its environment.
+//!
 //! [geth]: https://geth.ethereum.org/
 
 use {
+    dashmap::DashMap,
     std::{process::Stdio, sync::Arc},
-    tokio::sync::Mutex,
 };
 
 const PORT: &str = "8547";
-const GETH_PORT: &str = "8545";
 const GETH_PROGRAM: &str = "geth";
-const GETH_ARGS: [&str; 8] = [
+const GETH_ARGS: [&str; 6] = [
     "--dev",
     "--http",
     "--http.addr",
     "0.0.0.0",
     "--http.api",
     "web3,eth,net,debug",
-    "--http.port",
-    GETH_PORT,
 ];
 
+/// The base datadir used to ensure that the genesis block is always the same.
+const BASE_DATADIR: &str = "/base-datadir";
+
 #[tokio::main]
 async fn main() {
     run().await
 }
 
 // Avoid the IDE issues caused by #[tokio::main].
 async fn run() {
-    let child = Geth::start().await;
-
     let router: axum::Router<State> = axum::Router::new();
-    let router = router.route("/", axum::routing::post(restart));
-    let router = router.with_state(State(Arc::new(child)));
+    let router = router.route("/", axum::routing::post(start));
+    let router = router.route("/:port", axum::routing::delete(stop));
+    let router = router.with_state(State(Arc::new(Default::default())));
 
     axum::Server::bind(&format!("0.0.0.0:{PORT}").parse().unwrap())
         .serve(router.into_make_service())
         .await
         .unwrap();
 }
 
-#[derive(Debug)]
-struct Geth {
-    child: Mutex<tokio::process::Child>,
-}
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct Port(String);
 
-impl Geth {
-    async fn start() -> Self {
-        Self {
-            child: Mutex::new(Self::spawn().await),
-        }
+impl Port {
+    fn free() -> Self {
+        Self(
+            std::net::TcpListener::bind("0.0.0.0:0")
+                .unwrap()
+                .local_addr()
+                .unwrap()
+                .port()
+                .to_string(),
+        )
     }
+}
 
-    async fn restart(&self) {
-        let mut child = self.child.lock().await;
-        child.kill().await.unwrap();
-        *child = Self::spawn().await;
-    }
+#[derive(Debug, Default)]
+struct Processes(DashMap<Port, tokio::process::Child>);
 
-    async fn spawn() -> tokio::process::Child {
-        let child = tokio::process::Command::new(dbg!(GETH_PROGRAM))
-            .args(GETH_ARGS)
+impl Processes {
+    async fn start(&self) -> Port {
+        let port = Port::free();
```

Comment:
This is a race condition. `Port::free` only holds onto the port while TcpListener is alive. Afterwards the port can be reused. We should hold a mutex during `Processes::start`. In practice the current code is unlikely to go wrong because linux counts the port number up and this is the only thing running in the container but would still be better completely correct.

Reply:
Even without this problem, I feel Mutex<HashMap> would be better than DashMap. We discussed this in another PR before. IMO Mutex is simpler than DashMap. I would only use DashMap when it is important for performance that there is no lock. This is just a nit pick and doesn't block the PR but I wanted to mention it again.

Reply:
> This is a race condition.

Point taken - I was aware of this but figured it's not worth worrying about since this is just testing setup and the chance that this causes a problem is really low.

> We should hold a mutex during Processes::start

Holding a mutex during `Process::start` is a bottleneck because now you can't start another geth instance until the previous one has been bound and responsive and this can take a few seconds. So now instead of having the setup for four tests take a few seconds, the setup takes four * a few seconds, so you lose the parallelism.

I even thought that maybe better than trying to bind to `0.0.0.0:0` would be to just generate a random port and hope that it's free and accept the chance that it isn't.

The only "perfect" solution is to start geth with `0.0.0.0:0` and parse the logging output to try to figure out the port, but obviously parsing logging output is extremely sketchy and prone to error later down the line if the format changes.

> linux counts the port number up

I'm not sure, it seems to just be handing out random ports when I run it :shrug:

Reply:
If starting geth takes a while until it actually binds to the port then I agree that we shouldn't mutex there. In that case let's add a comment to this code mentioning the race condition and that it is unlikely to happen.

Maybe another reasonable solution is that you keep track of port numbers yourself. If you know that the docker container only runs this rust program then you can have a global atomic int next_port  that keeps track of the next port you want to give out.

Reply:
> Maybe another reasonable solution is that you keep track of port numbers yourself. If you know that the docker container only runs this rust program then you can have a global atomic int next_port that keeps track of the next port you want to give out

That's a good idea, I could allocate a fixed range of ports since this is the only program running in the container.

Reply:
https://github.com/cowprotocol/services/pull/1121/commits/ea95df09eed6a3ffc4161d776351b2132bf45874